WearTilesKotlin/app/src/main/java/com/example/wear/tiles/messaging/MessagingTileService.kt

```diff
@@ -69,181 +45,68 @@ private const val ID_CONTACT_PREFIX = "contact_"
-                            .setOnClick(ActionBuilders.LoadAction.Builder().build())
-                            .build()
-                    )
-                )
-                .addContent(Spacer.Builder().setWidth(SPACING_CONTACTS_HORIZONTAL).build())
-                .addContent(
-                    contactLayout(
-                        contact = contacts[2],
-                        clickable = Clickable.Builder()
-                            .setOnClick(ActionBuilders.LoadAction.Builder().build())
-                            .build()
-                    )
-                )
-                .build()
-        )
-        .addContent(Spacer.Builder().setHeight(SPACING_CONTACTS_VERTICAL).build())
-        .addContent(
-            Row.Builder()
-                .addContent(
-                    contactLayout(
-                        contact = contacts[3],
-                        clickable = Clickable.Builder()
-                            .setOnClick(ActionBuilders.LoadAction.Builder().build())
-                            .build()
+    private suspend fun buildState(contacts: List<Contact>): MessagingTileState {
+        val avatars = coroutineScope {
+            contacts.map { contact ->
+                async {
+                    val avatar = imageLoader.loadAvatar(
+                        context = this@MessagingTileService,
+                        contact = contact
                     )
-                )
-                .addContent(Spacer.Builder().setWidth(SPACING_CONTACTS_HORIZONTAL).build())
-                .addContent(searchLayout())
-                .build()
-        )
-        .setModifiers(
-            Modifiers.Builder()
-                .setSemantics(
-                    Semantics.Builder()
-                        .setContentDescription(getString(R.string.tile_messaging_label))
-                        .build()
-                )
-                .build()
-        )
-        .build()
+                    avatar?.run {
+                        contact.id to avatar
+                    }
+                }
+            }
+        }.awaitAll().filterNotNull().toMap()
+        return MessagingTileState(contacts, avatars)
+    }
 
-    private fun contactLayout(
-        contact: Contact,
-        clickable: Clickable
-    ) = Button.Builder(this, clickable).apply {
-        setContentDescription(contact.name)
-        setButtonColors(
-            ButtonColors(
-                ContextCompat.getColor(baseContext, R.color.secondary),
-                ContextCompat.getColor(baseContext, R.color.primary)
-            )
-        )
-        if (contact.avatarRes == null) {
-            setTextContent(contact.initials)
-        } else {
-            setImageContent("$ID_CONTACT_PREFIX${contact.id}")
+    override suspend fun tileRequest(requestParams: TileRequest): Tile {
+        val tileState = tileStateFlow.filterNotNull().first()
+
+        if (tileState.contacts.isEmpty()) {
+            updateContacts()
```

Comment:
I stupidly made is an async operation, I'll fix that.

```
    private fun updateContacts() {
        lifecycleScope.launch {
```

Reply:
forceUpdates probably shouldn't be called on tileStateFlow, because it assumes the tile service is running.

I think ideally you don't have any routine tile expiry, just push a tile update when the data validly changes.

I'll change to call the update externally when we update the data.  But there isn't something clearly driving that, but I can add it and slightly change the data.

Reply:
This probably deserves a more thorough look.  Is the StateFlow useful here, or should apps just read directly from the DB (Pref DataStore) in a one shot query?

Want me to take another stab, or would you like to take a look.

Regardless, I think this sample serves as a more realistic example, so glad we are having these discussions.

Reply:
I don't thoughts either way here yet - happy that this serves as a more realistic example, but let's iterate (at a later date).